### PR TITLE
added data storage

### PR DESCRIPTION
**Pull request by Flaivers**

### CODE REVIEW

File: src/browserStorage.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { Student } from "./student";
+import { IStorage } from "./interface";
+export class BaseBrowserStorage implements IStorage {
```
Comment (03leg) [inline src/browserStorage.ts]: rename pls all files, className === fileName (for example, BaseBrowserStorage -> BaseBrowserStorage.ts)

Comment (03leg) [inline src/browserStorage.ts]: why isn't class abstract? please rewatch our video session

File: src/browserStorage.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      return Promise.resolve(result);
+    }
+    return Promise.resolve([]);
+  }
+
+  saveStudents(students: Student[]): void {
+    this.getStorage().setItem(this._key, JSON.stringify(students));
+  }
+
+  getStorage() {
```
Comment (03leg) [inline src/browserStorage.ts]: why isn't method abstract?

File: src/localStorage.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Student } from "./student";
+import { BaseBrowserStorage } from "./browserStorage";
+export class StudentsLocalStorage extends BaseBrowserStorage {
+  getStudents(): Promise<Student[]> {
```
Comment (03leg) [inline src/localStorage.ts]: For what are methods `getStudents` and `saveStudents` here? why don't you use them from base class?

File: src/studentManager.ts [hunk 4]
```diff
@@ -1,8 +1,22 @@
...
+  constructor(typeStorage: IStorage) {
+    this._storage = typeStorage;
+  }
+  private _students = [new Student("Bob", "Ann", 2000, 1)];
+  get students(): Student[] {
+    return this._students;
+  }
 
-  students = [new Student("Bob", "Ann", 2000, 1)];
+  private _storage: IStorage = new StudentsLocalStorage(); //localStorage
```
Comment (03leg) [inline src/studentManager.ts]: plz remove initialization with `StudentsLocalStorage` here. `_storage` type is `IStorage | null` and default value is 'null' then create private property `storage` of type `IStorage` where if the _storage is null throw error. And after use `storage` instead `_storage` in this class.

**Review by 03leg — CHANGES_REQUESTED** [review]

File: src/browserStorage.ts (hunk 1 above)
Comment (Flaivers) [reply src/browserStorage.ts]: done

Comment (Flaivers) [reply src/browserStorage.ts]: done

File: src/browserStorage.ts (hunk 2 above)
Comment (Flaivers) [reply src/browserStorage.ts]: done

File: src/localStorage.ts (hunk 3 above)
Comment (Flaivers) [reply src/localStorage.ts]: done

File: src/studentManager.ts (hunk 4 above)
Comment (Flaivers) [reply src/studentManager.ts]: done

File: src/studentManager.ts [hunk 5]
```diff
@@ -1,8 +1,25 @@
...
+  constructor(typeStorage: IStorage) {
+    this._storage = typeStorage;
+  }
+  private _students = [new Student("Bob", "Ann", 2000, 1)];
+  get students(): Student[] {
+    return this._students;
+  }
+  
+  private _storage: IStorage | null = null;
+  private storage(): IStorage {
```
Comment (03leg) [inline src/studentManager.ts]: it's method, not property

**Review by 03leg — CHANGES_REQUESTED** [review]

**Review by 03leg — APPROVED** [review]

**Pull request merged**